### PR TITLE
StorageFile class: load method should not create a new file #129

### LINKED ISSUE #129: StorageFile class: load method should not create a new file
The current `load` method creates an empty file if the file is missing. 
1. This is not mentioned in the header comment.
2. Saving should not be the job of the `load` method in the first place.

### PR DESCRIPTION
**Pull request by yamgent**
Fixes #129.

```
load() method creates a new address book xml file if it
does not exist.

However, the responsibility of load() is to simply read the file,
not write it. Even if the file does not exist, load() should simply
treat the situation as if it is loading a blank file, and handle
accordingly. Creating a new file is redundant anyway if load() does
not even attempt to read it after creation, and load() will not gain
any new information by reading it.

Let's remove the saving functionality of load() in order to narrow
down the responsibility of the method to purely just reading the xml
files.
```

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@yamgent submitted v1 for review.

* [v1 1/3] [Move handling of missing file to normal code path](https://github.com/se-edu/addressbook-level2/pull/142/commits/2c76ccfb064f930c42784dd40a06b419bd7dc1ef)
* [v1 2/3] [Remove creating new file on load if not found ](https://github.com/se-edu/addressbook-level2/pull/142/commits/964bcd17b0725510e4d7a329319f43e3ed470e65)
* [v1 3/3] [Update javadocs for load() ](https://github.com/se-edu/addressbook-level2/pull/142/commits/349f9accc25f9c2825429a9af6b71ff93da84db1)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/07adaeece6cd84655dce171eb3a04828a4108c0e...349f9accc25f9c2825429a9af6b71ff93da84db1))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/142/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by damithc** [comment]
Oops, didn't realize this is reviewer-less :-p @yamgent can invite 2 peer reviewers?

File: src/seedu/addressbook/storage/StorageFile.java [hunk 1]
```diff
@@ -128,17 +138,9 @@ public AddressBook load() throws StorageOperationException {
...
-         * situation (i.e. not truly exceptional) we should not use an exception to handle it.
-         */
-
-        // create empty file if not found
         } catch (FileNotFoundException fnfe) {
-            final AddressBook empty = new AddressBook();
-            save(empty);
-            return empty;
-
+            assert false : "not possible: is already handled earlier";
```
Comment (MightyCupcakes) [inline src/seedu/addressbook/storage/StorageFile.java]: Since there is a specific exception for file not found, why not handle the logic here (that is reutrn the empty addressbook here) instead of adding an extra check above?

Comment (yamgent) [inline src/seedu/addressbook/storage/StorageFile.java]: - Should not change the path of execution
- Unit test

File: src/seedu/addressbook/storage/StorageFile.java [hunk 2]
```diff
@@ -113,12 +114,21 @@ public void save(AddressBook addressBook) throws StorageOperationException {
 
     /**
      * Loads data from this storage file.
-     *
+     * 
+     * @return an {@link AddressBook} containing the data in the file, or an empty {@link AddressBook} if it
```
Comment (Zhiyuan-Amos) [inline src/seedu/addressbook/storage/StorageFile.java]: Maybe can reword it to `@return the existing {@link AddressBook} if it exists. Else, returns an empty {@link AddressBook}`

File: src/seedu/addressbook/storage/StorageFile.java [hunk 3]
```diff
@@ -113,12 +114,21 @@ public void save(AddressBook addressBook) throws StorageOperationException {
...
+     *    does not exist.
      * @throws StorageOperationException if there were errors reading and/or converting data from file.
      */
     public AddressBook load() throws StorageOperationException {
+        File fileOnDisk = path.toFile();
+
+        if (!fileOnDisk.exists()) {
+            final AddressBook empty = new AddressBook();
```
Comment (Zhiyuan-Amos) [inline src/seedu/addressbook/storage/StorageFile.java]: Change the name from `empty` to `emptyAddressBook`. Otherwise, just delete this line and `return new AddressBook()` since it seems rather intuitive that `new AddressBook()` is an empty one

File: src/seedu/addressbook/storage/StorageFile.java (hunk 1 above)
Comment (Zhiyuan-Amos) [inline src/seedu/addressbook/storage/StorageFile.java]: Yes, we should not handle the logic here in exceptions, since file is missing / file does not exist is a normal situation i.e when user first uses this application.

Can we reword the  `assert false` sentence though. This sounds a bit weird ><

File: src/seedu/addressbook/storage/StorageFile.java (hunk 2 above)
Comment (MightyCupcakes) [inline src/seedu/addressbook/storage/StorageFile.java]: Don't think that wording is apporpriate in this context. load() is called during the program start up where there is no existing AddressBook object. It passes in the file with all the saved data and the method should return a new AddressBook object with all the previously saved data in it which is what the header comment is trying to imply.

Comment (Zhiyuan-Amos) [inline src/seedu/addressbook/storage/StorageFile.java]: Ah I see.

File: src/seedu/addressbook/storage/StorageFile.java (hunk 1 above)
Comment (pyokagan) [inline src/seedu/addressbook/storage/StorageFile.java]: Note that we are moving towards replacing all of our `assert false` with `throw new AssertionError()`

Comment (pyokagan) [inline src/seedu/addressbook/storage/StorageFile.java]: > Note that we are moving towards replacing all of our assert false with throw new AssertionError()

Oh sorry, didn't look at the full diff :-P Well, just FYI anyway.

**Comment by CanIHasReview** [comment]
# v2

@yamgent submitted v2 for review.

* [v2 1/3] [Move handling of missing file to normal code path](https://github.com/se-edu/addressbook-level2/pull/142/commits/16ef215339fcfd6531fe4285b6a961c329d0d62e)
* [v2 2/3] [Remove creating new file on load\(\)](https://github.com/se-edu/addressbook-level2/pull/142/commits/628d6c308291eb6dfd6b327e358415a0a8465f65)
* [v2 3/3] [Add unit test for loading non\-existant files](https://github.com/se-edu/addressbook-level2/pull/142/commits/fe2c1c65b1272fbd8c820fae0bcbe3e96520b4bf)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/f92b1d0567e30a36f32d8bd03eae2f3c466fa767...fe2c1c65b1272fbd8c820fae0bcbe3e96520b4bf)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level2/commit/db8919b1a06f13441dc021a33011190eb4d2947d))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/142/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/seedu/addressbook/storage/StorageFile.java [hunk 4]
```diff
@@ -117,8 +118,17 @@ public void save(AddressBook addressBook) throws StorageOperationException {
      * @throws StorageOperationException if there were errors reading and/or converting data from file.
      */
     public AddressBook load() throws StorageOperationException {
+        File fileOnDisk = path.toFile();
+
+        // create empty file if not found
+        if (!fileOnDisk.exists()) {
```
Comment (MightyCupcakes) [inline src/seedu/addressbook/storage/StorageFile.java]: It is not sufficient to check if the specified file exists or not. What if I create a new folder called "addressbook.xml"?

File: src/seedu/addressbook/storage/StorageFile.java [hunk 5]
```diff
@@ -139,7 +138,7 @@ public AddressBook load() throws StorageOperationException {
             return loaded.toModelType();
 
         } catch (FileNotFoundException fnfe) {
-            assert false : "not possible: is already handled earlier";
+            assert false : "A non-existant file scenario is already handled earlier.";
```
Comment (MightyCupcakes) [inline src/seedu/addressbook/storage/StorageFile.java]: see [#264](https://github.com/se-edu/addressbook-level4/issues/264)

File: src/seedu/addressbook/storage/StorageFile.java [hunk 6]
```diff
@@ -115,16 +115,15 @@ public void save(AddressBook addressBook) throws StorageOperationException {
...
      *
+     * @return an {@link AddressBook} containing the data in the file, or an empty {@link AddressBook} if it
+     *    does not exist.
      * @throws StorageOperationException if there were errors reading and/or converting data from file.
      */
     public AddressBook load() throws StorageOperationException {
         File fileOnDisk = path.toFile();
 
-        // create empty file if not found
         if (!fileOnDisk.exists()) {
```
Comment (MightyCupcakes) [inline src/seedu/addressbook/storage/StorageFile.java]: See my previous comment about this line which got overwritten

File: src/seedu/addressbook/storage/StorageFile.java (hunk 4 above)
Comment (yamgent) [inline src/seedu/addressbook/storage/StorageFile.java]: Thanks for spotting that! I will replace this with a more rigid method.

**Comment by CanIHasReview** [comment]
# v3

@yamgent submitted v3 for review.

* [v3 1/3] [Move handling of missing file to normal code path](https://github.com/se-edu/addressbook-level2/pull/142/commits/0916c0c85b1e915bb0c58f30b174b87aeea16d1b)
* [v3 2/3] [Remove creating new file on load\(\)](https://github.com/se-edu/addressbook-level2/pull/142/commits/c92d7f317f5f0c4f9375c1dcf0dc6355c15adb30)
* [v3 3/3] [Add unit test for loading non\-existant files](https://github.com/se-edu/addressbook-level2/pull/142/commits/a94ba5a97e0e768673e37e95d2a2d2bfd990bdfe)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/f92b1d0567e30a36f32d8bd03eae2f3c466fa767...a94ba5a97e0e768673e37e95d2a2d2bfd990bdfe)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level2/commit/78adda36abb3103bedf5ad2d2ab785efddc05ee3))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/142/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: test/java/seedu/addressbook/storage/StorageFileTest.java [hunk 7]
```diff
@@ -60,6 +64,15 @@ public void load_validFormat() throws Exception {
     }
 
     @Test
+    public void load_nonExistantFile_returnsEmptyAddressBook() throws Exception {
+        AddressBook actualAB = getStorage(NON_EXISTANT_FILE_NAME).load();
+        AddressBook expectedAB = new AddressBook();
+        
+        assertEquals(actualAB, expectedAB);
+        assertFileDoesNotExist(TEST_DATA_FOLDER + "/" + NON_EXISTANT_FILE_NAME);
```
Comment (damithc) [inline test/java/seedu/addressbook/storage/StorageFileTest.java]: Add a comment to explain that we are verifying the file is not created during loading. Otherwise the reader might wonder why this verification is done at this point.

File: test/java/seedu/addressbook/util/TestUtil.java [hunk 8]
```diff
@@ -137,4 +139,11 @@ public static void assertTextFilesEqual(Path path1, Path path2) throws IOExcepti
         List<String> list2 = Files.readAllLines(path2, Charset.defaultCharset());
         assertEquals(String.join("\n", list1), String.join("\n", list2));
     }
+
+    /**
+     * Asserts that the file given does not exist on the filesystem.
+     */
+    public static void assertFileDoesNotExist(String filePath) {
+        assertTrue(Files.notExists(Paths.get(filePath)));
```
Comment (MightyCupcakes) [inline test/java/seedu/addressbook/util/TestUtil.java]: Is there really a need to put this in TestUtil? Is there a rationale behind this? Because it looks to me it will only be used once.

Comment (damithc) [inline test/java/seedu/addressbook/util/TestUtil.java]: It's best to put potentially reusable functions in Util classes. That way, when someone needs a similar function in future, they can look through those Util classes to see if it already exists, instead of going through the entire code base. 

As a general rule, any code that is not tightly coupled to the class in which it originated can be generalised and pushed out of the class to a more suitable location. That way, our classes can remain lean and cohesive while maximising reuse.

**Comment by CanIHasReview** [comment]
# v4

@yamgent submitted v4 for review.

* [v4 1/3] [Move handling of missing file to normal code path](https://github.com/se-edu/addressbook-level2/pull/142/commits/0916c0c85b1e915bb0c58f30b174b87aeea16d1b)
* [v4 2/3] [Remove creating new file on load\(\)](https://github.com/se-edu/addressbook-level2/pull/142/commits/c92d7f317f5f0c4f9375c1dcf0dc6355c15adb30)
* [v4 3/3] [Add unit test for loading non\-existant files](https://github.com/se-edu/addressbook-level2/pull/142/commits/b0b8acc2ca26578aedb182b4e2435ddcb7e8ac86)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/f92b1d0567e30a36f32d8bd03eae2f3c466fa767...b0b8acc2ca26578aedb182b4e2435ddcb7e8ac86)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level2/commit/1e4f4bd3827b8af11bfb179c1671108d6eadf09e))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/142/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/seedu/addressbook/storage/StorageFile.java [hunk 9]
```diff
@@ -128,17 +136,8 @@ public AddressBook load() throws StorageOperationException {
...
-         * situation (i.e. not truly exceptional) we should not use an exception to handle it.
-         */
-
-        // create empty file if not found
         } catch (FileNotFoundException fnfe) {
-            final AddressBook empty = new AddressBook();
-            save(empty);
-            return empty;
-
+            throw new AssertionError("A non-existant file scenario is already handled earlier.");
```
Comment (kychua) [inline src/seedu/addressbook/storage/StorageFile.java]: `non-existant` -> `non-existent`

**Comment by CanIHasReview** [comment]
# v5

@yamgent submitted v5 for review.

* [v5 1/3] [Move handling of missing file to normal code path](https://github.com/se-edu/addressbook-level2/pull/142/commits/2ef44b21f5fda84c2e250ece0cc233483247612a)
* [v5 2/3] [Remove creating new file on load\(\)](https://github.com/se-edu/addressbook-level2/pull/142/commits/44ceadb348c32d4569aefd96c44babb6106331d0)
* [v5 3/3] [Add unit test for loading non\-existent files](https://github.com/se-edu/addressbook-level2/pull/142/commits/adf4aefa2a7b911a2c0bc163d1db9b25da2e5fa2)

([:books: Archive](https://github.com/se-edu/addressbook-level2/compare/5810f7c47b3f5392b75184bf95a896df33993744...adf4aefa2a7b911a2c0bc163d1db9b25da2e5fa2)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level2/compare/d25d0371d014b7a7fd0ac68d7d3834cf80e0a932~1...d25d0371d014b7a7fd0ac68d7d3834cf80e0a932))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level2.git refs/pr/142/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by yamgent** [comment]
May I know the status of this PR?

**Comment by damithc** [comment]
While we wait one more day to see if @pyokagan has any further comments, may be @yamgent can propose a commit message body for the merge commit? You can insert that in the PR description, like so:

```
commit message body (high level WHAT and WHY aspect of the PR -- but not HOW aspect)
```

Proposing merge commit message body is not part of the workflow yet, but we can trial it in a few PRs first.

**Comment by yamgent** [comment]
@damithc Edited the PR description for a proposed merge commit message body.

**Comment by damithc** [comment]
> @damithc Edited the PR description for a proposed merge commit message body.

I formatted it as code (to show how it will appear in commit message and removed the word 'currently'.
Remind me to merge this if there are no further comments over the weekend.
